### PR TITLE
Text rendering optimization

### PR DESCRIPTION
**Pull request by kategimranova**
Added char2array (dict with numpy arrays for all symbols) for faster text rendering

### CODE REVIEW

File: src/main.py [hunk 1]
```diff
@@ -16,6 +16,8 @@
 from src.utils.config import TransformerConfig, TrainingConfig, VTRConfig
 from src.utils.train import train
 
+import pickle
```
Comment (SpirinEgor) [inline src/main.py:19]: Please, use correct order of imports 😅 PyCharm can do it for you: [link](https://www.jetbrains.com/pycharm/guide/tips/optimize-imports/)

File: src/main.py [hunk 2]
```diff
@@ -104,7 +106,9 @@ def train_vtr_encoder(args: Namespace, train_data: list, val_data: list = None,
         not args.no_ocr,
     )
 
-    dataset_args = (vtr.font, vtr.font_size, vtr.window_size, vtr.stride, training_config.max_seq_len)
+    char2array = pickle.load(open(vtr.char2array, "rb"))
```
Comment (SpirinEgor) [inline src/main.py:109]: This way the file will not be closed. You need to use `with` context

File: src/main.py [hunk 3]
```diff
@@ -156,7 +160,9 @@ def train_vtr_encoder_sl(args: Namespace, train_data: list, val_data: list = Non
     )
     criterion = BceLossForTokenClassification()
 
-    dataset_args = (vtr.font, vtr.font_size, vtr.window_size, vtr.stride, training_config.max_seq_len)
+    char2array = pickle.load(open(vtr.char2array, "rb"))
```
Comment (SpirinEgor) [inline src/main.py:163]: Same issue

File: src/utils/config.py [hunk 4]
```diff
@@ -14,6 +14,7 @@ class VTRConfig:
     hidden_size_ocr: int
     num_layers_ocr: int
     ratio: float
+    char2array: str
```
Comment (SpirinEgor) [inline src/utils/config.py:17]: See `tokenizer` argument in `configure_arg_parser` in `main.py`. Let's do the same.

File: src/utils/slicer.py [hunk 5]
```diff
@@ -9,21 +9,16 @@
 class VTRSlicerWithText:
     def __init__(
         self,
-        window_size: int = 25,
-        stride: int = 10,
-        font: str = "fonts/NotoSans.ttf",
-        font_size: int = 15,
+        char2array: dict,
+        window_size: int = 32,
+        stride: int = 5,
```
Comment (SpirinEgor) [inline src/utils/slicer.py:14]: why did you change default values?

File: src/utils/slicer.py [hunk 6]
```diff
@@ -34,9 +29,11 @@ def __call__(self, text: str, max_slice_count: int = None) -> tuple[torch.Tensor
         if len(text) == 0:
             text = " "
 
+        unknown_token = self.char2array["UNK"]
```
Comment (SpirinEgor) [inline src/utils/slicer.py:32]: Move this to `__init__`

File: src/utils/slicer.py [hunk 7]
```diff
@@ -75,29 +71,26 @@ def __call__(self, text: str, max_slice_count: int = None) -> tuple[torch.Tensor
...
         self.stride = stride
-        self.font = font
-        self.font_size = font_size
+        self.char2array = char2array
 
     def __call__(self, text: str, max_slice_count: int = None) -> torch.Tensor:
-        image = text2image(text, font=self.font, font_size=self.font_size)
-        image_bytes = torch.as_tensor(np.array(image))
+        unknown_token = self.char2array["UNK"]
+        image_bytes = torch.Tensor(np.concatenate([self.char2array.get(char, unknown_token) for char in text], axis=1))
```
Comment (SpirinEgor) [inline src/utils/slicer.py:85]: `Tensor` is a type and not for direct usage, use `tensor` function to create it (https://pytorch.org/docs/stable/tensors.html#torch.Tensor)

**Review by SpirinEgor — CHANGES_REQUESTED** [review]

File: src/utils/slicer.py (hunk 5 above)
Comment (kategimranova) [reply src/utils/slicer.py:14]: Because these default values are in parser arguments: https://github.com/deepvk/vitrina/blob/8247cf637b717a4f48abe0e5b238dd6719962923/src/utils/config.py#L32
https://github.com/deepvk/vitrina/blob/8247cf637b717a4f48abe0e5b238dd6719962923/src/utils/config.py#L33
Should I need to return old values?

Comment (SpirinEgor) [reply src/utils/slicer.py:14]: Got it. Let's keep these values)

File: src/utils/config.py (hunk 4 above)
Comment (kategimranova) [reply src/utils/config.py:17]: Did I understand correctly that I should move `char2array` parameter from `VTRConfig` to `main` arg parser?

Comment (SpirinEgor) [reply src/utils/config.py:17]: Yes!

**Review by SpirinEgor — APPROVED** [review]
👍
